crates/shared/src/remaining_amounts.rs

```diff
@@ -48,19 +36,80 @@ impl Remaining {
                 order.metadata.executed_sell_amount_before_fees,
             ),
         };
+
         if order.data.partially_fillable {
-            Self::from_partially_fillable(total, executed)
+            let execution = Ratio::new_raw(
+                total
+                    .checked_sub(executed)
+                    .context("executed larger than total")?,
+                total,
+            );
+
+            let sell_amount = ratio::scalar_mul(&execution, order.data.sell_amount)
+                .context("overflow scaling sell amount for execution")?;
+            let fee_amount = ratio::scalar_mul(&execution, order.data.fee_amount)
+                .context("overflow scaling fee amount for execution")?;
+
+            let need = sell_amount
+                .checked_add(fee_amount)
+                .context("partially fillable need calculation overflow")?;
+
+            let balance = if sell_balance < need {
+                Ratio::new_raw(sell_balance, need)
+            } else {
+                ratio::one()
+            };
+
+            Ok(Self { execution, balance })
         } else {
-            Ok(Self::from_fill_or_kill(!executed.is_zero()))
+            let sell = order
+                .data
+                .sell_amount
+                .checked_add(order.data.fee_amount)
+                .context("overflow sell + fee amount")?;
+
+            let execution = if executed.is_zero() {
+                ratio::one()
+            } else {
+                ratio::zero()
+            };
+
+            let balance = if sell_balance >= sell {
+                ratio::one()
+            } else {
+                ratio::zero()
+            };
+
+            Ok(Self { execution, balance })
         }
     }
 
     /// Returns Err if the contract would error due to intermediate overflow.
     pub fn remaining(&self, total: U256) -> Result<U256> {
-        total
-            .checked_mul(self.numerator)
-            .and_then(|product| product.checked_div(self.denominator))
-            .context("overflow")
+        ratio::scalar_mul(
+            &self.balance,
+            ratio::scalar_mul(&self.execution, total)
+                .context("overflow scaling for order execution")?,
+        )
+        .context("overflow scaling for available balance")
```

Comment:
The smart contract does not concern itself with the balance of the user for the scaling so I think this should probably not throw on overflows, right?
In that case the scaling based on balance could be done with `Ratio<U512>` and only error if the end result is greater than `U256::MAX`.

Reply:
Why do you give me so much work @MartinquaXD  😛. Yes, you are right - this should not error on overflows in intermediate calculations. Will patch.